Donations by candidate

SchaeStewart opened the pull request:
Modified an route to return candidate details.  
Added a new route to return contributor+contribution data for a given `sboe_id`

closes #57

benjaminEwhite commented on server/api_doc.md:
```diff
@@ -100,15 +100,12 @@ response:
-- `offset` default value `0`
-- `limit` default value `50`
-
 Example: `/api/candidate/STA-C0498N-C-002?limit=2`
```
`limit` should go away in the example query as no longer supported by endpoint

benjaminEwhite commented on server/index.js:
```diff
@@ -71,13 +70,46 @@ api.get('/candidate/:ncsbeID', async (req, res) => {
+
+    client = await getClient()
+    const contributions = await client.query(
+      `select *, count(*) over () as full_count
```
Some of the fields returned are ambiguous since this is combining contributor and contributions data. Consider prefixing the fields that come from the contributor table with `contributor_` so it's clear — especially for the `id` field that's coming out.  So something like


```json
{
      "source_contribution_id": "962e55dc-14f3-4bc3-b7f6-6cd39441b3b8",
      "contributor_id": "7f731531-2ee1-4225-928e-91b11e50a56f",
      "transaction_type": "Individual",
      "committee_name": "COOPER FOR NORTH CAROLINA",
      "committee_sboe_id": "STA-C0498N-C-002",
      "committee_street_1": "434 FAYETTEVILLE ST   SUITE 2020",
      "committee_street_2": null,
      "committee_city": "RALEIGH",
      "committee_state": "NC",
      "committee_zip_code": "27601",
      "report_name": "2019 Year End Semi-Annual",
      "date_occurred": "9/13/19",
      "account_code": "Not Available",
      "amount": 1000,
      "form_of_payment": "Credit Card",
      "purpose": null,
      "candidate_or_referendum_name": null,
      "declaration": null,
      "contributor_id": "7f731531-2ee1-4225-928e-91b11e50a56f",
      "contributor_name": "Martha Armstrong",
      "contributor_city": "Wilmington",
      "contributor_state": "NC",
      "contributor_zip_code": "28411-8375",
      "contributor_profession": "Self Employed",
      "contributor_employer_name": "Consultant",
      "full_count": "25059"
    },
```

SchaeStewart replied:
I can do that, but those field names actually align with the field names from the data we get from the board of elections, and when you join the `contributor` and `contributions` tables you essentially get a copy of the `raw_contributions`.

I also think it might be better to keep our field names consistent so I'm reluctant to change it here, but I'm willing to if everyone thinks it's a good idea.

Also, in your example you end up with two `contributor_id` fields. The `id` field on `contributors` table is the `contributor_id` field on the `contributions` table.

Just giving my thoughts, I could be persuaded either way. If the team thinks prefixing is better, I'm happy to do it.

@Dafondo @benjaminEwhite

Dafondo replied:
That's a good point. I think it's okay to keep the field names as-is, if we document the field definitions clearly in the API doc.

benjaminEwhite replied:
Can get on board with keeping field names as is.

benjaminEwhite replied:
Can get on board for keeping names as is if that's consensus view. 

One thing to consider is do we really need to include the committee fields. This API resource is about contributions to a candidate, which doesn't conceptually imply committees — in fact, since ncsbeId is what it matches on, the committee data is going to be the same for all entries. To my mind would be cleaner to only include the data about the contribution.

SchaeStewart replied:
Yeah, that makes sense. I'll exclude the committee data from the results

SchaeStewart commented:
@benjaminEwhite I removed the committee info (except ncsbe id) from the response. Also, I just realized that I fixed the `/api/candidates/:ncsbeID` problem in this PR already

Dafondo commented on server/api_doc.md:
```diff
@@ -100,16 +100,13 @@ response:
 
 The `:ncsbeID` parameter must be URIEncoded (ie: `encodeURIComponent`)
 
 Query Params (optional):
```
No query params for this route. Can be removed?

benjaminEwhite commented on server/index.js:
```diff
@@ -72,13 +71,65 @@ api.get('/candidate/:ncsbeID', async (req, res) => {
+      [ncsbeID]
+    )
+    return res.send({
+      data: contributions.rows,
```
Query looks right at line 74, but should call it `committee` or `candidate` at line 73, and then update line 81 accordingly (just to avoid confusion).

benjaminEwhite replied:
Instead of having `data: [...]`, might consider having `data: {...}`. This will only ever return a single item, and conceptually, the endpoint suggests it returns info on a single person.

benjaminEwhite commented on server/index.js:
```diff
@@ -72,13 +71,65 @@ api.get('/candidate/:ncsbeID', async (req, res) => {
-      order by c.date_occurred asc
-      limit $2
-      offset $3
+      limit 1
```
given that sboe_id is the primary key on the committees table, this should only ever return 1, so no need for limit.

SchaeStewart commented:
@benjaminEwhite fixed

benjaminEwhite commented:
Merge away @SchaeStewart
